middleware/grpc_instrumentation.go

```diff
@@ -173,18 +171,115 @@ func (s *instrumentedClientStream) CloseSend() error {
+		opt(&instrumentationLabel)
+	}
+	return instrumentationLabel
+}
+
+type InstrumentationLabel struct {
+	acceptGRPCStatuses bool
+	maskHTTPStatuses   bool
+}
+
+func (i *InstrumentationLabel) AcceptGRPCStatuses(acceptance bool) {
+	i.acceptGRPCStatuses = acceptance
+}
+
+// GetInstrumentationLabel converts an error into an error code string by applying the configurations
+// contained in this InstrumentationLabel object.
+func (i *InstrumentationLabel) GetInstrumentationLabel(err error) string {
+	statusCode := i.errorToStatusCode(err)
+	return i.statusCodeToString(statusCode)
+}
+
+// statusCodeToString converts a given status code in its string representation.
+//
+//   - If the given status code is codes.OK, statusCodeToString returns "success" or "2xx",
+//     depending on whether maskHTTPStatuses is set to false or true respectively.
+//   - If the given status code is codes.Canceled, statusCodeToString returns "cancel"
+//     independently of maskHTTPStatuses.
+//   - If the given status code ia a valid HTTP status code, statusCodeToString returns the
+//     string representation of the HTTP status code or the first digit of the HTTP status
+//     code followed by "xx", depending on whether maskHTTPStatuses is set to false or true
+//     respectively.
+//   - If the given status code is a gRPC status code different from codes.Unknown, statusCodeToString
+//     returns the string representation of the status code, independently of maskHTTPStatuses.
+//   - If the given status code is codes.Unknown, statusCodeToString returns "error".
+func (i *InstrumentationLabel) statusCodeToString(statusCode codes.Code) string {
+	if statusCode == codes.OK {
+		if i.maskHTTPStatuses {
+			return "2xx"
+		}
+		return "success"
+	}
+
+	if statusCode == codes.Canceled {
 		return "cancel"
-	} else {
+	}
+
+	if statusCode == codes.Unknown {
 		return "error"
 	}
+
+	if isHTTPStatusCode(statusCode) {
+		statusFamily := int(statusCode / 100)
+		if i.maskHTTPStatuses {
+			return strconv.Itoa(statusFamily) + "xx"
+		}
+		return strconv.Itoa(int(statusCode))
+	}
+
+	return statusCode.String()
+}
+
+// errorToStatusCode extracts a status code from the given error, and does the following:
+//
+//   - If the error corresponds to context.Canceled, codes.Canceled is returned.
+//   - If the extracted status code is a valid HTTP status code, it is returned.
+//   - If the extracted status code is a gRPC code, and acceptGRPCStatusCodes is
+//     true, the gRPC status code is returned. Otherwise, codes.Unknown is returned.
+func (i *InstrumentationLabel) errorToStatusCode(err error) codes.Code {
+	if err == nil {
+		return codes.OK
+	}
+
+	if errors.Is(err, context.Canceled) {
+		return codes.Canceled
+	}
+
+	statusCode := grpcutil.ErrorToStatusCode(err)
+
+	if statusCode == codes.Canceled {
+		return statusCode
+	}
+
+	if isHTTPStatusCode(statusCode) {
+		return statusCode
+	}
+	if i.acceptGRPCStatuses {
+		return statusCode
+	}
+	return codes.Unknown
```

Comment:
I think this function is doing too much -- it should just return `grpcutil.ErrorToStatusCode(err)` for non-nil and not-canceled errors. (It also doesn't need to be a method of `InstrumentationLabel`.)

Reply:
This function is the only place where we actually take into account whether `i.acceptGRPCStatues` is set to true or not. If we move it out from here, we need to put it in `statusCodeToString()` function. I can try to see how it would look like, but it will for sure complicate the test.

Reply:
> I can try to see how it would look like, but it will for sure complicate the test.

I'm not sure if test is too complicated or not, but I think this function looks good now. Thank you for giving it a try.